fc/parsing/actions.py

```diff
@@ -873,19 +874,80 @@ def lhs_to_sympy(self, variable_generator, number_generator):
+
+    def to_sympy(self, variable_generator, number_generator):
+        """Convert this equation to Sympy.
+
+        Typically this will be called via :meth:`ModelInterface.convert_equations_to_sympy` which sets up appropriate
+        generators.
+
+        :param variable_generator: a function to resolve a name reference within the equation to a model variable.
+        :param number_generator: converts a number with units to a Sympy entity.
+        """
+        index = self.indexing_variable.to_sympy(variable_generator, number_generator)
+        # Load the data from file
+        data_file_name = os.path.basename(self.data_path)
+        if not os.path.exists(self.data_path):
+            raise ProtocolError(f'Unable to load data file {self.data_path} for interpolation.')
+        data = numpy.loadtxt(self.data_path, dtype=float, delimiter=',', ndmin=2, unpack=True)
+        assert data.ndim == 2
+        if data.shape[0] != 2:
+            raise ProtocolError(
+                f'The data file for an interpolate() must have 2 columns; file {data_file_name} has {data.shape[0]}')
+        # Check to see whether we can do a special-case optimisation for equally spaced independent variable points
+        steps = numpy.diff(data[0])
+        if abs(numpy.max(steps) - numpy.min(steps)) < 1e-6 * numpy.max(steps):
+            if steps[0] == 0.0:
+                raise ProtocolError(
+                    f'The data file {data_file_name} used in an interpolate() has all zero table steps.')
+            # Unit-convert the indexing variable if required
+            data_index_units = number_generator(1.0, self.index_units).units
+            index = index.model.convert_variable(
+                index, data_index_units, DataDirectionFlow.OUTPUT, move_annotations=False)
+            # Result is a special Symbol that can generate interpolation code
+            data_units = number_generator(1.0, self.data_units).units
+            return DataInterpolation(data_file_name, data, index, data_units)
+        else:
+            # Iterate over data rows to construct a piecewise representation for the interpolation
+            pieces = []
+            for row in range(data.shape[1] - 1):
+                if data[0, row] == data[0, row + 1]:
+                    # Vertical jumps are allowed in voltage clamp protocols, but trying to interpolate gives you a
+                    # divide by zero. Just skip the jump part, and we'll interpolate correctly at either side.
+                    continue
+                x1 = number_generator(data[0, row], self.index_units)
+                x2 = number_generator(data[0, row + 1], self.index_units)
+                y1 = number_generator(data[1, row], self.data_units)
+                y2 = number_generator(data[1, row + 1], self.data_units)
+                cond = sympy.And(index >= x1, index <= x2)
+                case = y1 + ((y2 - y1) / (x2 - x1)) * (index - x1)
+                pieces.append((case, cond))
```

Comment:
We should probably have something written in the FC manual about what happens to the left and right of the signal in the data file, and make sure this implementation does that?

Reply:
See the special case mentioned for `define` at https://chaste.cs.ox.ac.uk/trac/wiki/FunctionalCuration/ProtocolSyntax#Modelinterface

Reply:
I don't see where that specifies how extrapolation is handled?

Reply:
We're not extrapolating, only interpolating between the bounds of the piece condition.

Reply:
So what happens if the simulation goes beyond the bounds specified in the data? And what should happen? I think we need to define that

Reply:
Extrapolation is an error. Updated docs to say this.